### ISSUE TITLE
[Feature-Request] - Adding "Apply Button" GS Pluging settings window

### ISSUE THREAD

**Issue by MarcoEstevez** [PCSX2/pcsx2#794]
I always enjoy a lot do testing on PCSX2, Its exciting to see how things change almost every week.

Since changing settings are quite easy while games are currently running, could be really cool to have an extra Apply button, apart from "Ok" and "Cancel" on GS Panel.

I mean, "Apply" might do same as "Ok" but keeping plugin window visible, so then there is not need to open it again from main menu, I'm not sure how hard could that be, but for sure will make much easier adjust settings.

I also believe current windows size contain extra space to address a third button.

What do you think Guys?

**Comment by tsunami2311** [comment]
PCSX2 dont rendering when the gsdx window is open, so apply would be point less, unless they rewrite the GSDX window to no pause pcsx2

**Comment by MarcoEstevez** [comment]
I See, yes that's true. 

Then I guess there is no point with on here.

Thanks for answer.

**Comment by avih** [comment]
Well, there would be a point if the dialog would first be made to not stop emulation, and probably not modal too. It should be possible theoretically, though it's likely to require much attention to details of edge cases.

**Comment by MarcoEstevez** [comment]
So theoretically, if dialog were not modal anymore and emulator doesn't got paused, then it might not even need to have another "Apply" button.

I'm thinking now on commit every single change to the render flow. It could act like a real time control panel and then when you are done close the window by clicking "Ok".

Could that be possible?

**Comment by avih** [comment]
Most of the things could be possible. It's "only" a matter of a capable someone actually putting the time and effort to make it happen while not breaking things.

That being said, this specific suggestion, while with some merit, is both a philosophical choice and a practical one.

All the UI dialogs in PCSX2 apply their changes only after pressing the OK/Apply buttons, and there's value in allowing to change several values and then applying them together. Also, PCSX2 has to pause internally while applying any kind of change, and allowing rapid too changes could cause stability issues.

Overall, even if it can be made to work, I'd suggest not to do so before considering carefully the negative side effects from such change.

Regardless though, this cannot happen before the GSdx dislog is modified to not pause emulation, and that in itself might not be a trivial task.

**Comment by karasuhebi** [comment]
I know it won't be easy, but I 100% second this request. Being able to apply changes on the fly and/or being able to keep the GSdx Plugin Settings window open would be AMAZING.

**Comment by gregory38** [comment]
Which option do you want to change on the fly?

**Comment by karasuhebi** [comment]
IR, texture filtering, accurate blend, OGL depth, accurate DATE, AF and AA1. Under HW Hacks, I'd like MSAA, CRC Hacks and TC Offset.

**Comment by gregory38** [comment]
Ir/msaa/ogl depth requires to discard all textures. It is better to stop gsdx.

Accurate date can be changed on vsync.
Tf can be done on the fly.
Af likely on vsync as sampler must be recreated.
Accurate blending and crc hack could be done on the fly. It is better to wait the vsync to avoid strange behavior.

**Comment by gregory38** [comment]
Oh.aa1 and tex offset are safe.

**Comment by avih** [comment]
> Ir/msaa/ogl depth requires to discard all textures. It is better to stop gsdx.

Sure, but only when you click Apply ;)

There were two suggestions:

1. Add an Apply button to the dialog and don't hang emulation while the dialog is open. When Apply/OK are pressed, emulation will get paused/resumed as much as necessary and the changes will get applied. Clicking Apply will not close the dialog, clicking OK will close it.

2. Let the changes affect emulation even before clicking Apply/OK. This cannot be done before (1) is done.

For the sake of simplicity, let's focus this discussion on (1), and if it ever gets done, we can open a new issue to also discuss 2.

**Comment by vsub** [comment]
I always wanted to be able to switch between native and scaling mode with hotkey.

**Comment by karasuhebi** [comment]
I agree with @avih's suggestion. That's actually what I was thinking as well. Do 1 first and then 2 whenever you think you can tackle it, since that sounded like it was the hard part.

**Comment by avih** [comment]
> What about the stuff @gregory38 mentioned though? Would it be hard to make the "safe" changes be applied on vsync instead of waiting for the user to click "Apply"?

Didn't you just agree that we should not discuss 2 before 1 works?

**Comment by ssakash** [comment]
> Let the changes affect emulation even before clicking Apply/OK

If the changes do affect emulation before clicking Apply, then there is probably no reason for the Apply button to exist. Apply button is just pointless IMO.

**Comment by gregory38** [comment]
Discarding the texture is quite sensible. It must be done because it change the format of the target. Imagine the following situation.
1/ you render some image into a RT
2/ you reuse the RT as an input texture to draw additional element/effects.
What happen if you discard the RT between 1 and 2?

I think the best solution is to intercept the apply in the Vsync (i.e. between the rendering of 2 frames). And reapply the new change. Potentially with a texture flush but it is quite sensible. It is the pandora box.

I would suggest to start with the safe option. Tf/Af/blending/crc/some safe hack.

**Comment by avih** [comment]
> If the changes do affect emulation before clicking Apply, then there is probably no reason for the Apply button to exist. Apply button is just pointless IMO. 

But this still means to be able to open the dialog without pausing the emulation and applying changes without closing the dialog. In other words, (1).

So before (1) happens, there's no point in discussing (2), except theoretically (but there should be no discussion - it's possible), and it only takes the focus from a clear first step which has to be accomplished first.

**Comment by karasuhebi** [comment]
>Didn't you just agree that we should not discuss 2 before 1 works?

I agreed that because it was harder we should put it off, but if some of the safe ones are easy to make apply on-the-fly, why not? :) I get your point though, let's focus on 1 first. :D

**Comment by lightningterror** [comment]
Closing as it seems quite trivial, plus doesn't seem practical if you need to pause gs for some but not for other options.

**Comment by karasuhebi** [comment]
> Closing as it seems quite trivial, plus doesn't seem practical if you need to pause gs for some but not for other options.

I'm confused. One of the reasons it was closed was because it's trivial to do? Wouldn't that be a reason to _not_ close it?

Why doesn't it seem practical to allow GS to not pause when changing some options?
